File: sdk/swift/Sources/TelecomSDK/UsageAPI.swift
import Foundation

/// API for usage management
public class UsageAPI {
    private let client: HTTPClient
    
    public init(client: HTTPClient) {
        self.client = client
    }
    
    /// Get usage statistics for a subscriber
    public func getStats(subscriberID: Int64, startDate: Date, endDate: Date) async throws -> UsageStats {
        let formatter = ISO8601DateFormatter()
        let params: [String: String] = [
            "start_date": formatter.string(from: startDate),
            "end_date": formatter.string(from: endDate)
        ]
        
        return try await client.get(path: "/v1/subscribers/\(subscriberID)/usage", params: params)
    }
    
    /// List usage events
    public func listEvents(
        subscriberID: Int64? = nil,
        usageType: UsageType? = nil,
        startDate: Date? = nil,
        endDate: Date? = nil,
        page: Int = 1,
        pageSize: Int = 50
    ) async throws -> UsageEventList {
        var params: [String: String] = [
            "page": "\(page)",
            "page_size": "\(pageSize)"
        ]
        
        if let subscriberID = subscriberID {
            params["subscriber_id"] = "\(subscriberID)"
        }
        if let usageType = usageType {
            params["usage_type"] = "\(usageType)"
        }
        if let startDate = startDate {
            let formatter = ISO8601DateFormatter()
            params["start_date"] = formatter.string(from: startDate)
        }
        if let endDate = endDate {
            let formatter = ISO8601DateFormatter()
            params["end_date"] = formatter.string(from: endDate)
        }
        
        return try await client.get(path: "/v1/usage/events", params: params)
    }
    
    /// Get real-time usage for a subscriber
    public func getRealtime(subscriberID: Int64) async throws -> RealTimeUsage {
        return try await client.get(path: "/v1/subscribers/\(subscriberID)/realtime")
    }
}

File: sdk/kotlin/src/main/kotlin/com/telecom/UsageAPI.kt
package com.telecom

import java.time.format.DateTimeFormatter

/**
 * API for usage management
 */
class UsageAPI(private val client: HTTPClient) {
    
    private val formatter = DateTimeFormatter.ISO_DATE_TIME
    
    /**
     * Get usage statistics for a subscriber
     */
    suspend fun getStats(subscriberId: Long, startDate: java.time.LocalDateTime, endDate: java.time.LocalDateTime): UsageStats {
        val params = mapOf(
            "start_date" to formatter.format(startDate),
            "end_date" to formatter.format(endDate)
        )
        
        return client.get("/v1/subscribers/$subscriberId/usage", params)
    }
    
    /**
     * List usage events
     */
    suspend fun listEvents(
        subscriberId: Long? = null,
        usageType: UsageType? = null,
        startDate: java.time.LocalDateTime? = null,
        endDate: java.time.LocalDateTime? = null,
        page: Int = 1,
        pageSize: Int = 50
    ): UsageEventList {
        val params = mutableMapOf(
            "page" to page.toString(),
            "page_size" to pageSize.toString()
        )
        
        subscriberId?.let { params["subscriber_id"] = it.toString() }
        usageType?.let { params["usage_type"] = it.toString() }
        startDate?.let { params["start_date"] = formatter.format(it) }
        endDate?.let { params["end_date"] = formatter.format(it) }
        
        return client.get("/v1/usage/events", params)
    }
    
    /**
     * Get real-time usage for a subscriber
     */
    suspend fun getRealtime(subscriberId: Long): RealTimeUsage {
        return client.get("/v1/subscribers/$subscriberId/realtime")
    }
}
